databuilder/databuilder/models/type_metadata.py

```diff
@@ -0,0 +1,349 @@
+# Copyright Contributors to the Amundsen project.
+# SPDX-License-Identifier: Apache-2.0
+
+import abc
+from typing import Any, Dict, Iterator, Optional
+
+from databuilder.models.graph_node import GraphNode
+from databuilder.models.graph_relationship import GraphRelationship
+from databuilder.models.graph_serializable import GraphSerializable
+from databuilder.models.table_metadata import DescriptionMetadata
+
+
+class TypeMetadata(abc.ABC, GraphSerializable):
+    TYPE_NODE_LABEL = 'Column_Subtype'
+    TYPE_RELATION_TYPE = 'COLUMN_SUBTYPE'
+    INVERSE_TYPE_RELATION_TYPE = 'COLUMN_SUBTYPE_OF'
+    TYPE_KIND = 'kind'
+    TYPE_NAME = 'name'
+    TYPE_DESCRIPTION = 'description'
+    TYPE_DATA_TYPE = 'data_type'
+    TYPE_MAP_KEY = 'map_key'
+    TYPE_MAP_VALUE = 'map_value'
+    TYPE_SORT_ORDER = 'sort_order'
+
+    @abc.abstractmethod
+    def __init__(self,
+                 description: Optional[str] = None,
+                 start_label: Optional[str] = None,
+                 start_key: Optional[str] = None) -> None:
```

Comment:
so, not parent?

Reply:
re: the parent in the base class, after further thought and trying things out I'm not sure that it is worth it to have for the same reason I mentioned in a previous comment about why I didn't put the name in the base class

1. children of structs wouldn't know which `name` to use from their parent bc they would all have the same parent obj with a dict including all of them, so the benefit of being able to get the key in the base class wouldn't be available
2. if we are setting the child `data_type.parent` with the parent obj after the recursive bottom up children are created, we'd still need to be passing the start_label/start_key through the inits so it doesn't remove the need for that

I've added the parser with the recursive logic in my latest changes, check out the new stuff and we can discuss further

Reply:
base class:
```
def key(self):
  # key is always parent's key + own name
  parent_key = None
  if isinstance(self.parent, ColumnMetadata):
    parent_key = self.parent.get_column_key() # this doesn't exist atm
  else:
    parent_key = self.parent.key()
  return parent_key + '.' + self.name
```

Reply:
- let columns know their own key (maybe pass the table's key to each column metadata, for example)
- if an array has a child, that child's `name` will be a fake name, like `_inner_`. 
- `key()` can be implemented as above. Each node doesn't need to know it's own key, it only needs to know its own name and its parent and it can always get the key via logic above.